incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,148 @@
+
+    coalition_ids = range(number_of_coalitions)
+    all_metacoalitions = chain.from_iterable(
+        combinations(coalition_ids, size) for size in range(limit_of_revealed + 1)
+    )
+    all_metacoalition_ids = map(lambda x: Coalition.from_players(x).id, all_metacoalitions)
+    return np.fromiter(
+        all_metacoalition_ids, dtype=int,
+        # we can calculate the size of the resulting array using sum of the binomial coefficients for each size
+        count=coalitions_up_to(number_of_coalitions, limit_of_revealed))
+
+
+def coalitions_up_to(number_of_players: int, size_limit: int) -> int:
+    """Calculate the number of coalitions _smaller or equal_ to `size_limit`."""
+    return int(scipy.special.comb(number_of_players, np.arange(size_limit + 1)).sum())
+
+
+def get_coalition_player_id_map(number_of_players: int) -> np.ndarray:
+    """Map a coalition to its 'player_id'."""
+    viable_coalitions = [x.id for x in all_coalitions(number_of_players)
+                         if len(x) not in [0, 1, 2**number_of_players - 1]]
+    ret = np.zeros(2**number_of_players, dtype=int)
+    ret[viable_coalitions] = range(len(viable_coalitions))
+    return ret
+
+
+RMValue = np.float32
+
+
+class GameRegretMinimizer:
+    """The main regret minimizer object, holding the regret information.
+
+    We use the notion of "meta-coalitions" to refer to sets of coalitions.
+    This means that each coalition can be viewed as a player in a new game.
+    It would be wasteful to consider _all_ coalitions as players --
+    we don't ever need to consider coalitions which are already in the K_0 set.
+
+    That said, because we don't use all coalitions, we need to re-map their ids
+    (without that, we would have gaps in the ids, leading to inefficiencies).
+    To achieve this, we use a map from the coalitions (excluding K_0) to their "player_ids".
+
+    We minimize regret in every internal node of the game tree.
+    When updating regret, we need to scan through the meta-coalitions from largest to smallest.
+    This identification is refered to as the "rank" of the meta-coalition.
+    Such an order is found using the `metacoalition_ids_by_coalition_size` function, and stored
+    in `meta_id_to_rank`.
+
+    Attributes:
+        number_of_players: number of players in the games.
+        number_of_coalitions: number of coalitions which aren't in K_0.
+        limit_of_revealed: maximum number of coalitions to reveal, i.e. maximum size of metacoal.
+        meta_rank_to_id: map metacoalition size rank -> metacoalition id.
+        meta_id_to_rank: map metacoalition id -> metacoalition size rank.
+        regret: an array of instant regrets of each regret minimizer (zero if an action is inapplicable).
+            This array has size (number_of_regret_minimizers, number_of_coalitions).
+            It is indexed by the meta-coalition ids.
+    """
+
+    def __init__(self, number_of_players: int, limit_of_revealed: int) -> None:
```

Comment:
also needs `plus` and `predictive` boolean arguments.

Reply:
I'll add `predictive` in a separate issue. (https://github.com/furadnik/IncompleteCooperative/issues/346)